wrappers/services.go

```diff
@@ -66,37 +68,69 @@ func generateSnapServiceFile(app *snap.AppInfo) ([]byte, error) {
 	return genServiceFile(app), nil
 }
 
+func stopUserServices(cli *client.Client, inter interacter, services ...string) error {
+	ctx, cancel := context.WithTimeout(context.Background(), time.Duration(timeout.DefaultTimeout))
+	defer cancel()
+	failures, err := cli.ServicesStop(ctx, services)
```

Comment:
if I understand the code correctly failures can be non-empty even with err == nil?

should force returning an error if failures is non empty?  do we want to succeed a refresh that failed stopping services? I would start conservative but I'm open to input

Reply:
You are correct that err == nil if the session agent returns a well formed error response giving a list of command failures.

I am not entirely certain that propagating errors is the best choice here.  A mischievous user could replace their session agent with an alternative implementation that always returns errors: should that prevent the administrator from upgrading or removing snaps?

As for the `Kill` code path I skipped for user services: my understanding is that this was to work around bugs in old versions of systemd that have since been fixed.  The fact that this feature relies on a working user instance of systemd means we're already incompatible with those old versions.  So the existing session agent call-out should be sufficient to kill the service.  A second session agent API to hard kill services should not be necessary, and wouldn't offer snapd any reliability guarantees.

Reply:
@jhenstridge so if I read the code for system services correctly:

* we ignore but log failures to stop them?
* but error if fail to start them?

we should implement the same behavior for user services, at initiallly I think.  I'm not too worried about mischievous users, at least for now.

Reply:
I've adjusted the code to pass on errors from the session agent now (by altering the session agent client package to not suppress the error).